list.go

```diff
@@ -17,24 +17,132 @@ package hessian
 import (
 	"io"
 	"reflect"
+	"strings"
+	"time"
 )
 
 import (
 	perrors "github.com/pkg/errors"
 )
 
+var (
+	listTypeNameMapper = map[string]string{
```

Comment:
https://github.com/dubbogo/hessian2/blob/39ddacfb1012a468610e987d74eb9d5a88ccb5ee/object.go#L130-L137

As code above, `EncObject` may implicitly register the object, so I think that we'd better to add rwlock on listTypeNameMapper to avoid concurrent problem.

If we assume user will register the object at the first of his code explictly unless encoding or decodeing, then we can avoid rwlock. But in this situation, I think we should remove all implict registration codes.

Reply:
@gaoxinge  @fangyincheng I think implicit registration is needed to make it easy to use.

Reply:
@wongoo @fangyincheng If we allow implicit registration, we'd better to use `sync.Map` to avoid non thread safe problem.

Reply:
@wongoo I think this is also a tradeoff:

- If we only allow explicit registration at the first of user codes, we can 
   - write simpler codes, especially some reflect codes to extract type information from interface for registering
   - consider less concurrent problems
   - better performance because of less reflect and lock
   - may less confusion for user because user register type by himself
- If we use implicit registration, user can write simpler codes.

Reply:
When ` implicit registration`,  `v` need to implement pojo interface, then user will know it. 
Since user need implement pojo interface, why do not we ask user to register? 
If user don't need implement pojo interface, then I think ` implicit registration` would be best.

Reply:
@fangyincheng @gaoxinge I agree `explicit registration`, but it should do `implicit registration` for objects containing in the parameter `v` of `RegisterPOJO(v)`, include pojos, arrays, and multidimensional arrays.

Reply:
So I will add rwlock.

Reply:
@fangyincheng Why not try `sync.Map`? `sync.Map` may be simpler.

Reply:
@gaoxinge Doing.